src/nupic/algorithms/Cells4.cpp

```diff
@@ -1574,6 +1575,44 @@ void Cells4::applyGlobalDecay()
 
 //--------------------------------------------------------------------------------
 /**
+ * Helper function for Cells4::adaptSegment. Generates lists of synapses to
+ * decrement, increment, add, and remove.
+ *
+ * IMPLEMENTATION NOTE: The implementation makes a single pass through the
+ * containers for O(n) complexity.
+ *
+ */
+void Cells4::_generateListsOfSynapsesToAdjustForAdaptSegment(
+  const Segment& segment,
+  std::set<UInt>& synapsesSet,
+  std::vector<UInt>& inactiveSrcCellIdxs,
+  std::vector<UInt>& inactiveSynapseIdxs,
+  std::vector<UInt>& activeSrcCellIdxs,
+  std::vector<UInt>& activeSynapseIdxs)
+{
+  // Purge residual data
+  inactiveSrcCellIdxs.clear();
+  inactiveSynapseIdxs.clear();
+  activeSrcCellIdxs.clear();
+  activeSynapseIdxs.clear();
+
+  for (UInt i = 0; i != segment.size(); ++i) {
+    UInt srcCellIdx = segment[i].srcCellIdx();
+    if (not_in(srcCellIdx, synapsesSet)) {
+      inactiveSrcCellIdxs.push_back(srcCellIdx);  // TODO: Check synapse still exists!
```

Comment:
@scottpurdy, it's not my TODO. It came along for the ride when I extracted that code from `Cells4::adaptSegment`